--- tdebase-trinity-14.0.10/kicker/applets/clock/CMakeLists.txt.orig	2021-04-24 01:28:04.000000000 -0300
+++ tdebase-trinity-14.0.10/kicker/applets/clock/CMakeLists.txt	2021-05-10 19:56:35.734649946 -0300
@@ -45,6 +45,7 @@
 
 tde_add_library( ${target} STATIC_PIC AUTOMOC
   SOURCES ${${target}_SRCS}
+  DEPENDENCIES kickermain-shared
 )
 
 ##### clock_panelapplet (module) ################
@@ -57,7 +58,6 @@
 
 tde_add_kpart( ${target} AUTOMOC
   SOURCES ${${target}_SRCS}
-  DEPENDENCIES clock_panelapplet-static
   LINK kickermain-shared clock_panelapplet-static kickermain
   DESTINATION ${PLUGIN_INSTALL_DIR}
 )
