--- configure.ac	2022-01-13 09:23:53.000000000 +0000
+++ configure.ac	2022-07-20 11:59:07.503797192 +0100
@@ -415,7 +415,7 @@
 
 dnl Check for libssh and argp for SSH plugin
 if test "x$enable_ssh_token" = "xyes"; then
-	PKG_CHECK_MODULES([LIBSSH], [libssh])
+	PKG_CHECK_MODULES([LIBSSH], [libssh2])
 	AC_CHECK_DECLS([ssh_session_is_known_server], [], [], [#include <libssh/libssh.h>])
 	AC_CHECK_HEADER([argp.h], [], AC_MSG_ERROR([You need argp library.]))
 	saved_LIBS=$LIBS
