--- fittsen-20190618_277dd4f/fittsen.c.orig	2019-09-12 19:25:31.000000000 -0300
+++ fittsen-20190618_277dd4f/fittsen.c	2019-09-12 21:20:45.453103087 -0300
@@ -320,7 +320,7 @@
         return -1;
     }
 
-    char *post = "/.fittsen.cfg";
+    char *post = "/.config/fittsen/fittsenrc";
     char *fp = calloc(strlen(home)+strlen(post)+1, 1);
 
     if (fp == NULL) {
@@ -333,7 +333,7 @@
     strcat(fp, post);
 
     if (ini_parse(fp, handler, NULL) < 0) {
-        fprintf(stderr, "Cannot load ~/.fittsen.cfg\n");
+        fprintf(stderr, "Cannot load ~/.config/fittsen/fittsenrc\n");
         free(fp);
         return -1;
     }
