--- config.def.h.orig	2015-06-16 06:37:07.000000000 -0300
+++ config.def.h	2017-05-11 21:36:37.000000000 -0300
@@ -1,5 +1,5 @@
 /* See LICENSE file for copyright and license details. */
 
-static char *const rcinitcmd[] = { "/bin/rc.init", NULL };
-static char *const rcrebootcmd[] = { "/bin/rc.shutdown", "reboot", NULL };
-static char *const rcpoweroffcmd[] = { "/bin/rc.shutdown", "poweroff", NULL };
+static char *const rcinitcmd[] = { "/etc/rc.d/rc.startup", NULL };
+static char *const rcrebootcmd[] = { "/etc/rc.d/rc.shutdown", "reboot", NULL };
+static char *const rcpoweroffcmd[] = { "/etc/rc.d/rc.shutdown", "poweroff", NULL };
