--- projects/compiler-rt/cmake/config-ix.cmake.orig	2020-10-07 10:10:48.000000000 +0000
+++ projects/compiler-rt/cmake/config-ix.cmake	2021-01-04 10:30:37.088897130 +0000
@@ -624,7 +624,7 @@
     (OS_NAME MATCHES "Android|Darwin|Linux|FreeBSD|NetBSD|OpenBSD|Fuchsia|SunOS" OR
     (OS_NAME MATCHES "Windows" AND NOT CYGWIN AND
         (NOT MINGW OR CMAKE_CXX_COMPILER_ID MATCHES "Clang"))))
-  set(COMPILER_RT_HAS_SANITIZER_COMMON TRUE)
+  set(COMPILER_RT_HAS_SANITIZER_COMMON FALSE)
 else()
   set(COMPILER_RT_HAS_SANITIZER_COMMON FALSE)
 endif()
