--- Makefile.dist.orig	2017-01-31 16:09:25.000000000 -0300
+++ Makefile.dist	2017-02-01 09:21:14.000000000 -0300
@@ -30,8 +30,8 @@
 # number of ways. See the man page and the documentation in
 # doc/graft.{html,pdf,ps,txt}.
 
-PACKAGEDIR	= /usr/local/pkgs
-TARGETDIR	= /usr/local
+PACKAGEDIR	= /usr/pkg
+TARGETDIR	= /
 
 # You should only need to change the value of TOP to reflect the
 # installation directory for graft if it is different from your default
@@ -55,7 +55,7 @@
 #
 #	#!/usr/bin/perl -w
 
-PERL		= /usr/bin/perl
+PERL		= /tools/bin/perl
 
 # The location of the perl executable used to build the graft
 # executable. You'll need to change this only if you do not have a fixed
