--- tarlz-0.15/create.cc.orig	2019-04-09 17:30:13.000000000 -0300
+++ tarlz-0.15/create.cc	2019-07-23 14:53:21.004309339 -0300
@@ -30,9 +30,7 @@
 #include <unistd.h>
 #include <sys/stat.h>
 #include <sys/types.h>
-#if defined(__GNU_LIBRARY__)
 #include <sys/sysmacros.h>		// for major, minor
-#endif
 #include <ftw.h>
 #include <grp.h>
 #include <pwd.h>
--- tarlz-0.15/extract.cc.orig	2019-04-09 19:24:03.000000000 -0300
+++ tarlz-0.15/extract.cc	2019-07-23 14:53:38.000849445 -0300
@@ -32,9 +32,7 @@
 #include <utime.h>
 #include <sys/stat.h>
 #include <sys/types.h>
-#if defined(__GNU_LIBRARY__)
 #include <sys/sysmacros.h>		// for major, minor, makedev
-#endif
 #include <lzlib.h>
 
 #include "arg_parser.h"
