--- ca-certificates-20171114/update-ca.c.orig	2017-11-14 17:11:05.000000000 -0300
+++ ca-certificates-20171114/update-ca.c	2018-09-09 14:51:11.004815548 -0300
@@ -196,7 +196,7 @@
 
 static void proc_etccertsdir(const char* fullpath, struct hash* h, int tmpfile_fd)
 {
-	char linktarget[SYMLINK_MAX];
+	char linktarget[SYMLOOP_MAX];
 	ssize_t linklen;
 
 	linklen = readlink(fullpath, linktarget, sizeof(linktarget)-1);
