--- a/examples/logrotate-default	2017-05-04 14:28:48.393620994 -0400
+++ b/examples/logrotate-default	2017-05-04 14:48:58.138616506 -0400
@@ -12,9 +12,9 @@ create
 dateext
 
 # uncomment this if you want your log files compressed
-#compress
+compress
 
-# RPM packages drop log rotation information into this directory
+# Directory for application-specific configuration files
 include /etc/logrotate.d
 
 # no packages own wtmp and btmp -- we'll rotate them here
