--- lasagna/tain/tain_now.c.orig	2013-01-11 12:10:32.000000000 -0300
+++ lasagna/tain/tain_now.c	2017-05-18 23:45:24.000000000 -0300
@@ -13,11 +13,11 @@
 struct tain *
 tain_now(struct tain *t)
 {
-  struct timeval now;
+  struct timespec now;
 
-  gettimeofday(&now, NULL);
+  clock_gettime(CLOCK_REALTIME, &now);
   tain_load_utc(t, now.tv_sec);
-  t->nsec = (1000 * now.tv_usec) + 500;
+  t->nsec = now.tv_nsec;
 
   return t;
 }
