diff --git a/configure.ac b/configure.ac
index 18945fa8..41d25730 100644
--- a/configure.ac
+++ b/configure.ac
@@ -261,7 +261,7 @@ AC_TYPE_SIGNAL
 AC_FUNC_STRFTIME
 AC_FUNC_STRPTIME
 AC_FUNC_VPRINTF
-AC_CHECK_FUNCS([alarm btowc bzero ftime getcwd gettimeofday localtime_r memcmp memcpy memmove memset mkstemp munmap pstat_getdynamic raise re_comp regcomp select setlocale strchr strcspn strdup strerror strrchr strstr strtol strtoul timegm])
+AC_CHECK_FUNCS([alarm btowc bzero ftime getcwd gettimeofday localtime_r memcmp memcpy memmove memset mkstemp munmap pstat_getdynamic raise re_comp regcomp rresvport select setlocale strchr strcspn strdup strerror strrchr strstr strtol strtoul timegm])
 AC_REPLACE_FUNCS(snprintf vsnprintf)
 
 # More header checks--here use C++
diff --git a/hlnet/Connection.cc b/hlnet/Connection.cc
index fed71d68..51433b62 100644
--- a/hlnet/Connection.cc
+++ b/hlnet/Connection.cc
@@ -64,7 +64,7 @@
 
 typedef void (*SIGNAL_HANDLER) (...);
 
-#ifndef _MSC_VER                /* _WIN32 */
+#ifdef HAVE_RRESVPORT
 extern "C"
 {
   int rresvport (int *);
@@ -178,8 +178,7 @@ Connection::Open (int priv)
   {
     int aport = IPPORT_RESERVED - 1;
 
-//  Native Windows (MSVC) has no rresvport
-#ifndef _MSC_VER                /* _WIN32 */
+#ifdef HAVE_RRESVPORT
     sock = rresvport (&aport);
 #else
     return NOTOK;
