diff --git a/tdecore/network/tdesocketdevice.cpp b/tdecore/network/tdesocketdevice.cpp
index e4802bb7..b97ad0ba 100644
--- a/tdecore/network/tdesocketdevice.cpp
+++ b/tdecore/network/tdesocketdevice.cpp
@@ -39,7 +39,7 @@
 #include <unistd.h>
 
 #ifdef HAVE_POLL
-# include <sys/poll.h>
+# include <poll.h>
 #else
 # ifdef HAVE_SYS_SELECT
 #  include <sys/select.h>
diff --git a/tdeio/tdeio/connection.cpp b/tdeio/tdeio/connection.cpp
index fa316f56..9577ad53 100644
--- a/tdeio/tdeio/connection.cpp
+++ b/tdeio/tdeio/connection.cpp
@@ -27,7 +27,6 @@
 #include <tqtimer.h>
 
 #include <sys/types.h>
-#include <sys/signal.h>
 #include <sys/time.h>
 
 #include <errno.h>
