--- tdelibs-trinity-14.0.7/tdeio/kssl/kopenssl.cc.orig	2019-12-04 16:13:48.000000000 -0300
+++ tdelibs-trinity-14.0.7/tdeio/kssl/kopenssl.cc	2020-01-18 17:20:38.842247950 -0300
@@ -2048,7 +2048,7 @@
   return a->type;
 #endif
   kdWarning() << "X509_OBJECT_get_type not defined!" << endl;
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L && !defined(LIBRESSL_VERSION_NUMBER)
   return X509_LU_NONE;
 #else
   return 0;
@@ -2057,7 +2057,7 @@
 
 X509* KOpenSSLProxy::X509_OBJECT_get0_X509(const X509_OBJECT *a) {
   if (K_X509_OBJECT_get0_X509) return (K_X509_OBJECT_get0_X509)(a);
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
   return a->data.x509;
 #endif
   kdWarning() << "X509_OBJECT_get0_X509 not defined!" << endl;
@@ -2067,7 +2067,7 @@
 
 ASN1_TIME* KOpenSSLProxy::X509_getm_notAfter(const X509 *x) {
   if (K_X509_getm_notAfter) return (K_X509_getm_notAfter)(x);
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
   return X509_get_notAfter(x);
 #endif
   kdWarning() << "X509_get_notAfter not defined!" << endl;
@@ -2076,7 +2076,7 @@
 
 ASN1_TIME* KOpenSSLProxy::X509_getm_notBefore(const X509 *x) {
   if (K_X509_getm_notBefore) return (K_X509_getm_notBefore)(x);
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
   return X509_get_notBefore(x);
 #endif
   kdWarning() << "X509_get_notBefore not defined!" << endl;
