--- tdelibs-trinity-14.0.9/tdecore/CMakeLists.txt.orig	2020-10-16 20:20:38.000000000 -0300
+++ tdelibs-trinity-14.0.9/tdecore/CMakeLists.txt	2021-02-09 20:16:21.369220041 -0300
@@ -141,10 +141,10 @@
   SOURCES ${${target}_SRCS}
   VERSION 14.0.0
   EMBED tdecorenetwork-static ${TDEHW_LIB}
-  LINK DCOP-shared tdefx-shared ICE SM ${ZLIB_LIBRARIES}
-  LINK_PRIVATE ltdlc-static ${KDESVGICONS} ${XCOMPOSITE_LIBRARIES}
-    ${LIBIDN_LIBRARIES} ${LIBBFD_LIBRARIES} ${LIB_UTIL} ${GAMIN_LIBRARIES}
-    ${Backtrace_LIBRARY}
+  LINK ltdlc-static DCOP-shared tdefx-shared ICE SM ${ZLIB_LIBRARIES}
+   ${KDESVGICONS} ${XCOMPOSITE_LIBRARIES}
+  LINK_PRIVATE ${LIBIDN_LIBRARIES} ${LIBBFD_LIBRARIES} ${LIB_UTIL}
+   ${GAMIN_LIBRARIES} ${Backtrace_LIBRARY}
   DESTINATION ${LIB_INSTALL_DIR}
 )
 
