--- procps-ng-3.3.13/configure.ac.orig	2018-03-03 04:11:55.678869635 -0300
+++ procps-ng-3.3.13/configure.ac	2018-04-03 19:05:53.977065498 -0300
@@ -45,6 +45,8 @@
 # Checks for header files.
 AC_HEADER_MAJOR
 AC_CHECK_HEADERS([arpa/inet.h fcntl.h float.h langinfo.h libintl.h limits.h locale.h netinet/in.h stdint.h stdio_ext.h stdlib.h string.h sys/file.h sys/ioctl.h sys/param.h sys/time.h termios.h unistd.h utmp.h values.h wchar.h wctype.h])
+AC_CHECK_HEADERS([libio.h],
+[AC_DEFINE(HAVE_LIBIO_H, [1], [Define to 1 if libio.h present])], [])
 
 # Checks for typedefs, structures, and compiler characteristics.
 AC_CHECK_HEADER_STDBOOL
--- procps-ng-3.3.13/procio.c.orig	2018-03-03 03:59:18.064843718 -0300
+++ procps-ng-3.3.13/procio.c	2018-04-03 19:14:35.229074359 -0300
@@ -24,7 +24,9 @@
 #endif
 #include <errno.h>
 #include <fcntl.h>
-#include <libio.h>
+#if HAVE_LIBIO_H
+# include <libio.h>
+#endif
 #include <limits.h>
 #include <stdio.h>
 #include <stdlib.h>
