--- mdadm-4.1/mdadm.h.orig	2018-10-01 15:26:06.000000000 -0300
+++ mdadm-4.1/mdadm.h	2019-07-25 00:06:25.116883538 -0300
@@ -45,10 +45,7 @@
 #include	<errno.h>
 #include	<string.h>
 #include	<syslog.h>
-#ifdef __GLIBC__
-/* Newer glibc requires sys/sysmacros.h directly for makedev() */
 #include	<sys/sysmacros.h>
-#endif
 #ifdef __dietlibc__
 #include	<strings.h>
 /* dietlibc has deprecated random and srandom!! */
