--- linux-5.10.100/fs/reiser4/blocknrlist.c~	2022-02-14 20:42:57.012001822 -0300
+++ linux-5.10.100/fs/reiser4/blocknrlist.c	2022-02-14 20:43:49.381167073 -0300
@@ -109,7 +109,8 @@
  * ordering is to be preserved, @cmp must return 0."
  */
 static int blocknr_list_entry_compare(void* priv UNUSED_ARG,
-                                      struct list_head *a, struct list_head *b)
+				      const struct list_head *a,
+				      const struct list_head *b)
 {
 	blocknr_list_entry *entry_a, *entry_b;
 	reiser4_block_nr entry_a_end, entry_b_end;
