--- src/paths.h.orig	2018-03-28 13:49:50.000000000 -0300
+++ src/paths.h	2018-05-03 11:06:26.561140356 -0300
@@ -43,8 +43,8 @@
 #define INITLVL2	"/var/log/initrunlvl"	/* COMPAT: New runlevel */
 				/* Note: INITLVL2 definition needs INITLVL */
 #define HALTSCRIPT1	"/etc/init.d/halt"	/* Called by "fast" shutdown */
-#define HALTSCRIPT2	"/etc/rc.d/rc.0"	/* Called by "fast" shutdown */
+#define HALTSCRIPT2	"/etc/rc.d/rc.shutdown"	/* Called by "fast" shutdown */
 #define REBOOTSCRIPT1	"/etc/init.d/reboot"	/* Ditto. */
-#define REBOOTSCRIPT2	"/etc/rc.d/rc.6"	/* Ditto. */
+#define REBOOTSCRIPT2	"/etc/rc.d/rc.reboot"	/* Ditto. */
 #endif
 
